### PR TITLE
[FLINK-10325] [State TTL] Refactor TtlListState to use only loops, no java stream API for performance

### PR DESCRIPTION
**Pull request by azagrebin**
## What is the purpose of the change

Refactor TtlListState to use only loops instead of java stream API to exclude any performance impact

## Brief change log

  - Refactor TtlListState.updateTs() to use loop
  - Refactor TtlListState.collect() to use loop
  - auxiliary methods in TtlUtils

## Verifying this change

existing tests

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (yes)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Comment by Aitozi** [comment]
Hi @azagrebin ,does for loop perform better than stream api ?  And I have not found an existing test for the performance of `TtlListState`, may be we should add one to verify this change ?

**Comment by Clarkkkkk** [comment]
Hi @Aitozi, loop does perform better than stream api, but the gain by using loop seems so little with the operations like update and collect. However, for the simplicity, I prefer the  stream api. What do you think? @azagrebin

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/ttl/TtlListState.java [hunk 1]
```diff
@@ -73,11 +72,14 @@ public void addAll(List<T> values) throws Exception {
 		return () -> new IteratorWithCleanup(finalResult.iterator());
 	}
 
-	private void updateTs(List<TtlValue<T>> ttlValue) throws Exception {
-		List<TtlValue<T>> unexpiredWithUpdatedTs = ttlValue.stream()
-			.filter(v -> !expired(v))
-			.map(this::rewrapWithNewTs)
-			.collect(Collectors.toList());
+	private void updateTs(List<TtlValue<T>> ttlValues) throws Exception {
+		List<TtlValue<T>> unexpiredWithUpdatedTs = new ArrayList<>();
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/ttl/TtlListState.java]: Would it make sense to initialize the capacity of this `ArrayList` to `ttlValues.size`? That way we would save resizing work.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/ttl/TtlListState.java [hunk 2]
```diff
@@ -116,7 +125,12 @@ public void updateInternal(List<T> valueToStore) throws Exception {
 	}
 
 	private List<TtlValue<T>> withTs(List<T> values) {
-		return values.stream().map(this::wrapWithTs).collect(Collectors.toList());
+		List<TtlValue<T>> withTs = new ArrayList<>();
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/ttl/TtlListState.java]: Same here with initial capacity.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/ttl/TtlUtils.java [hunk 3]
```diff
@@ -21,11 +21,19 @@
 /** Common functions related to State TTL. */
 class TtlUtils {
 	static <V> boolean expired(TtlValue<V> ttlValue, long ttl, TtlTimeProvider timeProvider) {
-		return ttlValue != null && expired(ttlValue.getLastAccessTimestamp(), ttl, timeProvider);
+		return expired(ttlValue, ttl, timeProvider.currentTimestamp());
+	}
+
+	static <V> boolean expired(TtlValue<V> ttlValue, long ttl, long currentTimestamp) {
+		return ttlValue != null && expired(ttlValue.getLastAccessTimestamp(), ttl, currentTimestamp);
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/ttl/TtlUtils.java]: If `ttlValue` can be `null`, then we should annotate it with `@Nullable`.

**Comment by tillrohrmann** [comment]
@Clarkkkkk the `TtlListState` is somewhat performance critical and, thus, it makes sense imo to improve it as much as possible. One of the problems was also that we queried the timestamp for each element when updating it. With Andrey's changes this won't happen.

**Comment by Clarkkkkk** [comment]
@tillrohrmann Thanks for your reply. Sounds reasonable.

**Pull request closed without merging**